modules/efk-logging-fluentd-undefined.adoc

```diff
@@ -0,0 +1,121 @@
+and `CDM_EXTRA_KEEP_FIELDS` are not moved to `undefined`.
+|CDM_USE_UNDEFINED=true
+
+|CDM_UNDEFINED_NAME
+|Specify a name for the undefined top level field if using `CDM_USE_UNDEFINED`. The default is`undefined`.
+Enabled only when `CDM_USE_UNDEFINED` is `true`.
+|CDM_UNDEFINED_NAME="undef"
+
+|CDM_UNDEFINED_MAX_NUM_FIELDS
+|If the number of undefined fields is greater than this number, all undefined fields are converted to their JSON string representation 
+and stored in the `CDM_UNDEFINED_NAME` field. If the record contains more than this value of undefined fields, no further processing takes place on these fields. 
+Instead, the fields will be converted to a single string JSON value, stored in the top-level `CDM_UNDEFINED_NAME` field. Keeping the default of `-1`allows for an unlimited number of undefined fields, which is not recommended.
+Enabled only when `CDM_USE_UNDEFINED` is `true`.
+|CDM_UNDEFINED_MAX_NUM_FIELDS=4
+
+|CDM_UNDEFINED_TO_STRING
+|Set to `true` to convert all undefined fields to their JSON string representation. The default is `false`.
+|CDM_UNDEFINED_TO_STRING=true
+
+|CDM_UNDEFINED_DOT_REPLACE_CHAR
+|Specify a character to use in place of a dot character '.' in an undefined field. `MERGE_JSON_LOG` must be `true`. The default is `UNUSED`. 
+If you set the MERGE_JSON_LOG parameter to `true`, see the xref:mergejson[Note below].
+|CDM_UNDEFINED_DOT_REPLACE_CHAR="_"
+
+|===
+
+[[mergejson]]
+[NOTE]
+====
+If you set the `MERGE_JSON_LOG` parameter in the log collector daemonset and CDM_UNDEFINED_TO_STRING environment variables to true, you might receive an Elasticsearch *400* error. 
+The error occurs because when`MERGE_JSON_LOG=true`, the log collector adds fields with data types other than string. When you set CDM_UNDEFINED_TO_STRING=true, 
+the log collector attempts to add those fields as a string value resulting in the Elasticsearch 400 error. The error clears when the log collector rolls over 
+the indices for the next day’s logs
+
+When the log collector rolls over the indices, it creates a brand new index. The field definitions are updated and you will not get the *400* error.
+====
+
+.Procedure
+
+Use the `CDM_*` parameters to configure undefined and empty field processing.
+
+. Configure how to process defined fields, as needed:
+
+.. Specify the defined fields to move using the `CDM_DEFAULT_KEEP_FIELDS` and `CDM_EXTRA_KEEP_FIELDS`.
+
+.. Set `CDM_KEEP_EMPTY_FIELDS` to `true` to retain empty defined fields.
+
+. Configure how to process undefined fields, as needed:
+
+.. Set `CDM_USE_UNDEFINED` to `true` to move undefined fields:
+
+.. Specify a name for the undefined fields using the `CDM_UNDEFINED_NAME` parameter.
+
+.. Set `CDM_UNDEFINED_MAX_NUM_FIELDS` to a value other than the default `-1`, to set an upper bound on the number of undefined fields 
+in a single record.  
+
+. Set `UNDEFINED_TO_STRING` to `true` to convert undefined fields to JSON format. 
```

Comment:
The value is converted to the json string when set `UNDEFINED_TO_STRING` to `true`.

For example, if not set `UNDEFINED_TO_STRING` , we may see a field's value like:
 "Number": 10,
10 is an integer here
but if set `UNDEFINED_TO_STRING` to `true`, then the value should be:
"Number": "10",
"10" is a string here

Reply:
To be specific - `UNDEFINED_TO_STRING` will cause the value to be transformed to its JSON string representation.  For simple scalar values like numbers and booleans, this just means the value will be a quoted string - 10 -> "10", 3.1415 -> "3.1415", false -> "false".

For map/dict values and array values, they are converted to their JSON string representation: `"mapfield":{"key":"value"}` becomes `"mapfield":"{\"key\":\"value\"}"` and `"arrayfield":[1,2,"three"]` becomes `"arrayfield":"[1,2,\"three\"]"` - this at least preserves the structure of the value, so that you can retrieve the value later and convert it back to a map or an array.